### PR TITLE
Feature/issue 2428 all doc unmarshal rebased

### LINKED ISSUE #2428: Avoid document body unmarshalling on changes requests
Currently changes requests that specify style=all_docs have unnecessary performance overhead when retrieving branched documents.  

changes.addDocToChangeEntry will retrieve the document from the bucket (to identify the set of leaf revisions).  In the process it will unmarshal the entire doc.

Two enhancements:
 - Store the list of leaf rev ids in the revision cache along with the active revision, to avoid retrieving the document from the bucket when it's resident in the rev cache
 - When bucket retrieval is required, unmarshal only the rev history, not the full doc body.

### PR DESCRIPTION
**Pull request by tleyden**
Fixes #2428

Todo:

- [x] Run functional tests
    * http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-base-cc/439/
    * http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-base-di/404/
    * http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-topology-specific-di/320/
    * http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-topology-specific-cc/340/

### CODE REVIEW

**Comment by tleyden** [comment]
The DI functional tests are failing from what looks to be build errors:

http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-base-di/400/console

**Comment by tleyden** [comment]
The topo-specific DI tests looks like it got past the build and unit test phase:

http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-topology-specific-di/319/

**Comment by tleyden** [comment]
Ok I think I know what happened.  In http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-base-di/400/console it failed on the unit tests:

```
"--- FAIL: TestChangesActiveOnlyClock (7.65s)", "\tassert.go:31: ", "\tassert.go:31: ", "\tassert.go:31: 
--- FAIL: TestChangesActiveOnlyInteger (0.39s)\n\tassert.go:31: \n\tassert.go:31: \n\tassert.go:31: 
... etc
```

but in http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-topology-specific-di/319/ it didn't even run the unit tests.

**Comment by tleyden** [comment]
I'm getting a unit test failure locally with `TestChangesActiveOnlyInteger`:

```
	changes_api_test.go:1058
	assert.Equals(t, len(entry.Changes), 2)	

	expected: 2
	     got: 1
	assert.go:31: 
	assert.go:31: 
	assert.go:31: 

```

**Comment by tleyden** [comment]
I found the issue.  When initializing `documentRoot{}` via: 

```
docRoot := documentRoot{}
```

the `syncMeta.History` field ended up as empty.   Changing it to explicitly initialize the `History` with `make(RevTree)` (as is done in other parts of codebase), fixes the issue:

```
docRoot := documentRoot{
 		SyncData: &syncData{History: make(RevTree)},
}
```

The SG + Accel unit tests pass for me locally, and I've got a functional test re-run queued up at:

http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-base-di/402/

**Comment by tleyden** [comment]
Updated benchmark test results after fix for https://github.com/couchbase/sync_gateway/pull/2467#issuecomment-293727061:

| Fix Status  | Millisconds/Op | Bytes/Op | Allocations/Op |
| ------------- | ------------- | ------------- | ------------- |
| Without Fix  | 5074  | 699042928 | 1031780 | 
| Without Fix  | 5127  | 699030160 | 1031734 | 
| Without Fix  | 4689  | 699035600 | 1031753 | 
| With Fix  | 2238  | 431586592 | 514781 | 
| With Fix  | 2175  | 421275872 | 514373 | 
| With Fix  | 1983  | 409035872 | 513874 | 

Without fix: 8ef9a00e9154b03702f56aa401a3da1c1e2f6091
With fix: 44e34e9be5e586df1d89fadef63047327559dd7f

**Comment by tleyden** [comment]
@adamcfraser I think this is ready for review, pending functional tests passing.  A few are still in progress (see ticket description)

**Comment by tleyden** [comment]
>  A few are still in progress (see ticket description)

There were some failures, but I believe it's related to:

https://github.com/couchbaselabs/mobile-testkit/issues/1113

File: db/crud.go [hunk 1]
```diff
@@ -56,6 +56,29 @@ func (db *DatabaseContext) GetDoc(docid string) (*document, error) {
 	return doc, nil
 }
 
+func (db *DatabaseContext) GetDocSyncData(docid string) (syncData, error) {
```
Comment (adamcfraser) [inline db/crud.go]: When we merge this to master, we'll need to include some 'if db.UseXattrs' handling, similar to what we're doing in GetDoc.  Added a note on https://github.com/couchbase/sync_gateway/issues/2468.

**Comment by adamcfraser** [comment]
This looks good to me.

**Comment by tleyden** [comment]
All functional tests are passing now

File: db/crud.go (hunk 1 above)
Comment (tleyden) [reply db/crud.go]: Ok.  I'll add a TODO in the code

Comment (tleyden) [reply db/crud.go]: I just (force) pushed up an amended commit with a TODO and a comment on that function

**Pull request merged**